### PR TITLE
Fix DAC not working in Spring 

### LINKED ISSUE #30543: [BUG] DefaultAzureCredential can't work properly in Spring Cloud Azure
**Describe the bug**

Found this issue when fixing this doc https://github.com/MicrosoftDocs/azure-docs/pull/97305; the best case scenario is for users to run this application both locally and in Azure hosting environments without needing to modify any configurations.

If using `system-assigned managed identity`, users can use the following configuration
```properties
spring.cloud.azure.keyvault.secret.property-sources[0].endpoint=https://<your-keyvault-name>.vault.azure.net
```

But if users are using `user-assigned managed identity`, users can not make it work in the Azure hosting environment:
```properties
spring.cloud.azure.keyvault.secret.property-sources[0].endpoint=https://<your-keyvault-name>.vault.azure.net
spring.cloud.azure.keyvault.secret.property-sources[0].credential.client-id={Client ID of user-assigned managed identity}
```

The current solution I provided in the above PR is not ideal, because with such configurations, users won't be able to run the applications locally.
```properties
spring.cloud.azure.keyvault.secret.property-sources[0].endpoint=https://<your-keyvault-name>.vault.azure.net
spring.cloud.azure.keyvault.secret.property-sources[0].credential.managed-identity-enabled=true
```
or 
```properties
spring.cloud.azure.keyvault.secret.property-sources[0].endpoint=https://<your-keyvault-name>.vault.azure.net
spring.cloud.azure.keyvault.secret.property-sources[0].credential.managed-identity-enabled=true
spring.cloud.azure.keyvault.secret.property-sources[0].credential.client-id={Client ID of user-assigned managed identity}
```

***Exception or Stack Trace***
N/A

**To Reproduce**
Follow this https://github.com/MicrosoftDocs/azure-docs/pull/97305. 

***Code Snippet***
N/A

**Expected behavior**
With the following configurations, users can run the application both locally and in the Azure hosting environment (with user-assigned managed identity)
```properties
spring.cloud.azure.keyvault.secret.property-sources[0].endpoint=https://<your-keyvault-name>.vault.azure.net
spring.cloud.azure.keyvault.secret.property-sources[0].credential.client-id={Client ID of user-assigned managed identity}
```

**Screenshots**
N/A

**Setup (please complete the following information):**
 - OS: [e.g. MacOS]
 - IDE: [e.g. IntelliJ]
 - Library/Libraries: spring-cloud-azure-starter-keyvault-secrets:4.3.0
 - Java version: [e.g. 8] 
 - App Server/Environment: Azure Spring Apps
 - Frameworks: Spring Boot 

If you suspect a dependency version mismatch (e.g. you see `NoClassDefFoundError`, `NoSuchMethodError` or similar), please check out [Troubleshoot dependency version conflict article](https://aka.ms/azsdk/java/dependency/troubleshoot) first. If it doesn't provide solution for the problem, please provide:
- verbose dependency tree (`mvn dependency:tree -Dverbose`)
- exception message, full stack trace, and any available logs

**Additional context**
N/A

**Information Checklist**
Kindly make sure that you have added all the following information above and checkoff the required fields otherwise we will treat the issuer as an incomplete report
- [x] Bug Description Added
- [x] Repro Steps Added
- [x] Setup information Added

### PR DESCRIPTION
**Pull request by saragluna**
# Description

Fixes #30543. 

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

**Comment by saragluna** [comment]
@yiliuTo and @backwind1233, please help review this PR.

**Comment by backwind1233** [comment]
@saragluna I have a question
is it a valid scenario like this?

```
spring.cloud.azure.keyvault.secret.property-sources[0].endpoint=https://<your-keyvault-name-0>.vault.azure.net
spring.cloud.azure.keyvault.secret.property-sources[0].credential.client-id={Client ID-0 of user-assigned managed identity}
spring.cloud.azure.keyvault.secret.property-sources[1].endpoint=https://<your-keyvault-name-1>.vault.azure.net
spring.cloud.azure.keyvault.secret.property-sources[1].credential.client-id={Client ID-1 of user-assigned managed identity}

```

Will it work both local and Azure Env with different user assigned identities?

**Comment by saragluna** [comment]
> Hi @saragluna Xiaolu Dai FTE I have a question, is it a valid scenario like this?
> 
> ```
> spring.cloud.azure.keyvault.secret.property-sources[0].endpoint=https://<your-keyvault-name-0>.vault.azure.net
> spring.cloud.azure.keyvault.secret.property-sources[0].credential.client-id={Client ID-0 of user-assigned managed identity}
> spring.cloud.azure.keyvault.secret.property-sources[1].endpoint=https://<your-keyvault-name-1>.vault.azure.net
> spring.cloud.azure.keyvault.secret.property-sources[1].credential.client-id={Client ID-1 of user-assigned managed identity}
> ```
> 
> Will it work both local and Azure Env with different user assigned identities after current PR is merged?

Yes, it is valid. And yes, I think so.

**Comment by saragluna** [comment]
@backwind1233 and @yiliuTo could you help review this PR?

**Comment by backwind1233** [comment]
Do you think we need add some java doc?

File: sdk/spring/spring-cloud-azure-core/src/main/java/com/azure/spring/cloud/core/implementation/factory/AbstractAzureServiceClientBuilderFactory.java [hunk 1]
```diff
@@ -148,9 +143,8 @@ protected void configureCore(T builder) {
         configureConfiguration(builder);
```
Comment (backwind1233) [inline sdk/spring/spring-cloud-azure-core/src/main/java/com/azure/spring/cloud/core/implementation/factory/AbstractAzureServiceClientBuilderFactory.java:143]: image

I think we need to change the java doc here.

File: sdk/spring/spring-cloud-azure-core/src/test/java/com/azure/spring/cloud/core/implementation/credential/resolver/AzureTokenCredentialResolverTests.java [hunk 2]
```diff
@@ -45,15 +46,15 @@ void shouldResolveClientCertificateTokenCredential() {
...
+    void shouldResolveUserAssignedManagedIdentityTokenCredential() {
         AzureTestProperties properties = new AzureTestProperties();
         properties.getCredential().setClientId("test-mi-client-id");
         properties.getCredential().setManagedIdentityEnabled(true);
         Assertions.assertEquals(ManagedIdentityCredential.class, resolver.resolve(properties).getClass());
     }
 
     @Test
-    void shouldResolveSystemAssignedMITokenCredential() {
+    void shouldResolveSystemAssignedManagedIdentityTokenCredential() {
```
Comment (backwind1233) [inline sdk/spring/spring-cloud-azure-core/src/test/java/com/azure/spring/cloud/core/implementation/credential/resolver/AzureTokenCredentialResolverTests.java:57]: nice

**Review by backwind1233 — APPROVED** [review]
LGTM

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/context/AzureTokenCredentialAutoConfiguration.java [hunk 3]
```diff
@@ -62,14 +69,21 @@ public class AzureTokenCredentialAutoConfiguration extends AzureServiceConfigura
...
-        } else {
-            return factory.build().build();
-        }
+    TokenCredential tokenCredential(ClientSecretCredentialBuilderFactory clientSecretCredentialBuilderFactory,
+                                    ClientCertificateCredentialBuilderFactory clientCertificateCredentialBuilderFactory,
+                                    UsernamePasswordCredentialBuilderFactory usernamePasswordCredentialBuilderFactory,
+                                    ManagedIdentityCredentialBuilderFactory managedIdentityCredentialBuilderFactory,
+                                    DefaultAzureCredentialBuilderFactory defaultAzureCredentialBuilderFactory) {
+
+        List<AzureCredentialResolver<TokenCredential>> resolverList = new ArrayList<>();
```
Comment (stliu) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/context/AzureTokenCredentialAutoConfiguration.java:78]: I'm not keen to these bunch of **new**, it makes things not extendable

File: sdk/spring/spring-cloud-azure-core/src/main/java/com/azure/spring/cloud/core/implementation/credential/resolver/ClientCertificateCredentialResolver.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        String authorityHost = properties.getProfile().getEnvironment().getActiveDirectoryEndpoint();
+
+        if (authorityHost == null) {
+            authorityHost = AzureEnvironment.AZURE.getActiveDirectoryEndpoint();
+        }
+
+        ClientCertificateCredentialBuilderFactory factory = this.builderFactory == null
+            ? new ClientCertificateCredentialBuilderFactory(properties) : this.builderFactory;
+
+        ClientCertificateCredentialBuilder builder = factory
```
Comment (stliu) [inline sdk/spring/spring-cloud-azure-core/src/main/java/com/azure/spring/cloud/core/implementation/credential/resolver/ClientCertificateCredentialResolver.java:49]: this factory api looks very wired, it call `build()` before config the factory

File: sdk/spring/spring-cloud-azure-core/src/main/java/com/azure/spring/cloud/core/implementation/credential/resolver/DefaultAzureCredentialResolver.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        }
+
+        if (authorityHost == null) {
+            authorityHost = AzureEnvironment.AZURE.getActiveDirectoryEndpoint();
+        }
+
+        DefaultAzureCredentialBuilderFactory factory = this.builderFactory == null
+            ? new DefaultAzureCredentialBuilderFactory(properties) : this.builderFactory;
+
+        return factory
```
Comment (stliu) [inline sdk/spring/spring-cloud-azure-core/src/main/java/com/azure/spring/cloud/core/implementation/credential/resolver/DefaultAzureCredentialResolver.java:47]: this is wired too, it calls `build` twice

**Comment by saragluna** [comment]
Closing this PR now, will have a new design for it. https://github.com/Azure/azure-sdk-for-java/issues/32443

**Pull request closed without merging**